portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java

```diff
@@ -71,11 +76,19 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
         Date date = new Date();
         String email = credentials.getCurrentUser().getEmail();
 
+        Document doc =
+            Document.newBuilder().setContent(content).setType(Document.Type.PLAIN_TEXT).build();
+        LanguageServiceClient languageService = LanguageServiceClient.create();
+        Sentiment sentiment = languageService.analyzeSentiment(doc).getDocumentSentiment();
+        double score = sentiment.getScore();
+        languageService.close();
```

Comment:
Creating and closing the language service client for every request is probably not very efficient. You may want to store this on the servlet instance and use the same language service client for many requests.

note that in order to call `close`, you can override the servlet's `destroy` method, which is called once when the servlet engine is done with this servlet.

Reply:
What are the guidelines with using global variables across any given class? Is it typical or should we avoid? I've been told having global variables is bad style, but then again this (asynchronous?) programming seems to work non-linearly so there doesn't seem to be a nice way to avoid this, unless you want inefficiencies like here. 

Should I be overriding every servlet's delete method to close/delete global variables everywhere? Or is the space freed up/deleted automatically for most variables, and here I had to close because its some sort of streaming object?

Reply:
"Global" variables in Java refers to things stored in `static` fields. Things you declare in the servlet class *without* `static` are owned by a particular instance of the servlet. Servlets are run by a "servlet engine" which will create an instance of each servlet you've configured with `@WebServlet` (as if you'd called `new CommentServlet()`), then it will run `init()` on the servlet. Then every time a user makes a request to the server, it will call `doGet` and `doPost` appropriately. When you go to shut down the server, it will call `destroy()` on each servlet before shutting down.

Broadly, the problem(s) with global/shared variables is that storing data in a shared location can mean that you 'leak' data between multiple method calls unintentionally, and in multithreaded applications (like servers), you can have different threads writing the same data at the same type, which can cause data corruption, or cause one request to use data from a different request. This is what happened when you saved the user's email from `doGet` to use within `doPost`; requests from multiple users could happen at the same time, so that shared data would be shared between both users.

Using global, or just shared data (as is the case with fields in servlets) isn't inherently a bad thing. There are a lot of things in software, like database connectors and objects that provide *shareable* services which are written in thread-safe ways which are fine to declare in a shared location and use in many places. Some of these kinds of services may be declared as statics you can get through a static method, like `UserServiceFactory.getUserService()`. Others require you to explicitly initialized them and shut them down, like `languageService`, where the documentation specifically says that it creates threads and you should call `close` to shut down those threads.

You don't need to override destroy to shut down things which don't require it. Most objects don't create external resources that you would *need* to shut down, and its fine to just leave them be and let the garbage collector handle them.

Managing things that can be shared across requests is pretty much what the servlet `init` and `destroy` methods are for.